crates/ironrdp-rdpdr/src/pdu/efs.rs

```diff
@@ -0,0 +1,229 @@
+            0x554C => Ok(PacketId::PAKID_CORE_USER_LOGGEDON),
+            0x5543 => Ok(PacketId::PAKID_PRN_USING_XPS),
+            _ => Err(invalid_message_err!("try_from", "PacketId", "invalid value")),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum VersionAndIdPDUKind {
+    /// [2.2.2.2 Server Announce Request (DR_CORE_SERVER_ANNOUNCE_REQ)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/046047aa-62d8-49f9-bf16-7fe41880aaf4)
+    ServerAnnounceRequest,
+    /// [2.2.2.3 Client Announce Reply (DR_CORE_CLIENT_ANNOUNCE_RSP)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/d6fe6d1b-c145-4a6f-99aa-4fe3cdcea398)
+    ClientAnnounceReply,
+}
+
+#[derive(Debug)]
+/// VersionAndIdPDU is a fixed size structure representing multiple PDUs.
+/// The kind field is used to determine the actual PDU type.
+pub struct VersionAndIdPDU {
+    version_major: u16,
+    version_minor: u16,
+    pub client_id: u32,
+    kind: VersionAndIdPDUKind,
+}
+
+impl VersionAndIdPDU {
+    pub fn new(version_major: u16, version_minor: u16, client_id: u32, kind: VersionAndIdPDUKind) -> Self {
+        Self {
+            version_major,
+            version_minor,
+            client_id,
+            kind,
+        }
+    }
+
+    fn header(&self) -> PduResult<SharedHeader> {
+        match self.kind {
+            VersionAndIdPDUKind::ClientAnnounceReply => Ok(SharedHeader {
+                component: Component::RDPDR_CTYP_CORE,
+                packet_id: PacketId::PAKID_CORE_CLIENTID_CONFIRM,
+            }),
+            VersionAndIdPDUKind::ServerAnnounceRequest => Err(unexpected_message_type_err!("ServerAnnounceRequest", 0)),
+        }
+    }
+
+    pub fn decode(src: &mut ReadCursor, kind: VersionAndIdPDUKind) -> PduResult<Self> {
+        Ok(Self {
+            version_major: src.read_u16(),
+            version_minor: src.read_u16(),
+            client_id: src.read_u32(),
+            kind,
+        })
+    }
+}
+
+impl PduEncode for VersionAndIdPDU {
+    fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
+        self.header()?.encode(dst)?;
+        dst.write_u16(self.version_major);
+        dst.write_u16(self.version_minor);
+        dst.write_u32(self.client_id);
+        Ok(())
+    }
+
+    fn name(&self) -> &'static str {
+        match self.kind {
+            VersionAndIdPDUKind::ServerAnnounceRequest => "ServerAnnounceRequest",
+            VersionAndIdPDUKind::ClientAnnounceReply => "ClientAnnounceReply",
+        }
+    }
+
+    fn size(&self) -> usize {
+        // header bytes + (2 * u16) + u32 = header bytes + (2 * 2 bytes) + 4 bytes = header bytes + 8 bytes
+        self.header().unwrap().size() + 8
+    }
+}
+
+/// 2.2.2.4 Client Name Request (DR_CORE_CLIENT_NAME_REQ)
+/// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/902497f1-3b1c-4aee-95f8-1668f9b7b7d2
+#[derive(Debug)]
+pub struct ClientNameRequest {
```

Comment:
nit: Such structures could be transformed to be represented in mo rust-y way 😄 
```
pub enum ClientNameRequest {
    Ascii(String),
    Unicode(String),
}

impl ClientNameRequest {
    // - ClientNameRequestUnicodeFlag could be removed, flags could be directly transformed to/from enum variants
    // - `new` could be removed too
}
```

Reply:
34ec2d46c46c87c4dfb60537deff0f00b52536af